src/transformers/models/pop2piano/tokenization_pop2piano.py

```diff
@@ -0,0 +1,352 @@
+# coding=utf-8
+# Copyright 2023 The Pop2Piano Authors and The HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tokenization class for Pop2Piano."""
+
+import json
+import os
+from typing import Optional, Tuple, Union
+
+import numpy as np
+import pretty_midi
+import torch
+
+from ...tokenization_utils import AddedToken, PreTrainedTokenizer
+from ...utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+VOCAB_FILES_NAMES = {
+    "vocab_file": "vocab.json",
+}
+
+PRETRAINED_VOCAB_FILES_MAP = {
+    "vocab_file": {
+        "susnato/pop2piano_dev": "https://huggingface.co/susnato/pop2piano_dev/blob/main/vocab.json",
+    },
+}
+
+
+class Pop2PianoTokenizer(PreTrainedTokenizer):
+    """
+    Constructs a Pop2Piano tokenizer. This tokenizer does not require training. The `Pop2PianoTokenizer.__call__()`
+    method can only process one sequence at a time.
+
+    This tokenizer inherits from [`PreTrainedTokenizer`] which contains most of the main methods. Users should refer
+    to: this superclass for more information regarding those methods. However the code does not allow that and only
+    supports composing from various genres.
+
+    Args:
+        vocab_file (`str`):
+            Path to the tokenizer file which contains token-ids such as `TOKEN_SPECIAL`, `DEFAULT_VELOCITY`.
+        vocab_size_special (`int`, *optional*, defaults to 4):
+            Number of special values.
+        vocab_size_note (`int`, *optional*, defaults to 128):
+            This represents the number of Note Values. Note values indicate a pitch event for one of the MIDI pitches.
+            But only the 88 pitches corresponding to piano keys are actually used.
+        vocab_size_velocity (`int`, *optional*, defaults to 2):
+            Number of Velocity tokens.
+        vocab_size_time (`int`, *optional*, defaults to 100):
+            This represents the number of Beat Shifts. Beat Shift [100 values] Indicates the relative time shift within
+            the segment quantized into 8th-note beats(half-beats).
+    """
+
+    vocab_files_names = VOCAB_FILES_NAMES
+    pretrained_vocab_files_map = PRETRAINED_VOCAB_FILES_MAP
+
+    def __init__(
+        self,
+        vocab_file,
+        vocab_size_special: int = 4,
+        vocab_size_note: int = 128,
+        vocab_size_velocity: int = 2,
+        vocab_size_time: int = 100,
+        n_bars: int = 2,
```

Comment:
`num_bars`. Plus it's not documented. (Does the tokenizer actually use this?)

Reply:
yes it is used in `relative_batch_tokens_to_midi`.